File: 2D-World-Simulator-Python/OPP-Python/Organisms/Animals/Antelope.py
import random

from Game.Animal import Animal
from Game.OrganismStatics import *


class Antelope(Animal):
    def __init__(self, organism_world, point):
        super().__init__(organism_world, point, 4, 4, "A", OrganismSpecie.ANTELOPE, "#ffbf2d")

    # moving two fields instead of one
    def action(self):
        # getting new first point, and the second it should not be the initial position
        new_point1 = self.get_near_point(self.organism_position, False)
        new_point2 = self.get_near_point(new_point1, False)
        while True:
            if new_point2 != self.organism_position and new_point2 != new_point1:
                break
            new_point2 = self.get_near_point(new_point1, False)

        new_pos = self.organism_world.get_organism_at_pos(new_point2)

        # checking there is other organism
        if new_pos is not None:
            super().collision(new_pos)
        else:
            self.move_to_given_field(new_point2.get_x(), new_point2.get_y())

    # giving info about antelope defence success
    def defence(self, attacking_organism):
        if self.get_away(attacking_organism):
            return True
        return False

    # performing antelope defence function
    def get_away(self, attacking_organism):
        chance = random.randint(1, PROBABILITY_RANGE)
        if chance <= CHANCE_GETAWAY:
            # antelope runs away - moves to neighbouring free field
            new_point = self.get_near_point(self.organism_position, True)

            # no place to run, fight -> going back in collision function!
            if self.organism_position == new_point:
                return False
            # that means antelope defended itself and run away
            self.organism_world.get_world_window().add_comment(
                "ANTELOPE - > run away from " + attacking_organism.get_organism_sign())
            return True

        return False

File: 2D-World-Simulator-Java/OPP-Java/src/op/JavaProject/Organisms/Animals/Antelope.java
package op.JavaProject.Organisms.Animals;

import op.JavaProject.Functionality.BoardPoint;
import op.JavaProject.Functionality.Species;
import op.JavaProject.Game.Animal;
import op.JavaProject.Game.Organism;
import op.JavaProject.Game.World;

import java.awt.*;

import static op.JavaProject.Game.OrganismsStatics.CHANCE_GETAWAY;
import static op.JavaProject.Game.OrganismsStatics.PROBABILITY_RANGE;

public class Antelope extends Animal {

    public Antelope(World organismWorld, BoardPoint point) {
        super(organismWorld, point);
        this.strength = 4;
        this.initiative = 4;
        this.organismSign = "A";
        this.organismSpecie = Species.ANTELOPE;
        this.organismColor = new Color(255, 191, 45);
    }

    //moving two fields instead of one
    @Override
    public void Action() {
        BoardPoint initialPos = organismPos;
        //getting new first point
        BoardPoint newPoint1 = GetNearPoint(organismPos, false);

        //new second point shouldn't be the initial position
        BoardPoint newPoint2;
        do {
            newPoint2 = GetNearPoint(newPoint1, false);
        } while (newPoint2.equals(newPoint1) || newPoint2.equals(initialPos));

        Organism newPos = organismWorld.GetOrganismAtPos(newPoint2);

        //checking if there is other organism
        if (newPos != null) {
            Collision(newPos);
        } else {
            MoveToGivenField(newPoint2);
        }
    }

    //giving info about antelope defence success
    @Override
    public boolean Defence(Organism attackingOrganism) {
        if (GetAway(attackingOrganism)) return true;
        return false;
    }

    //performing antelope defence function
    private boolean GetAway(Organism attackingOrganism) {

        BoardPoint newPoint;
        int chance = organismPos.GetRandomValue(1, PROBABILITY_RANGE);
        if (chance <= CHANCE_GETAWAY) {
            //antelope runs away - moves to neighborhood free field
            newPoint = GetNearPoint(GetOrganismPos(), true);

            //no place to run, fight -> going back in collision function of parent class!
            if (GetOrganismPos().equals(newPoint)) return false;

            //that means antelope defended itself and run away
            organismWorld.GetCommentator().AddComment("ANTELOPE - > run away from " + attackingOrganism.GetOrganismSign());
            return true;
        }
        return false;
    }

}
